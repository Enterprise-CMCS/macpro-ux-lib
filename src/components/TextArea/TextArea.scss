@use "../../sass/macbis-colors" as c;

.usa-form-group .usa-textarea:focus {
  outline: 0.25rem solid c.$focus;
}
