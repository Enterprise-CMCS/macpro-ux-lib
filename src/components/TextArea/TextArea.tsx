import React, { forwardRef, useState } from "react";

type TextAreaElements = JSX.IntrinsicElements["textarea"];

interface Props extends TextAreaElements {
  id: string;
  label: string;
  name: string;
  characterCountMessage?: string;
  errorMessage?: string;
  inputError?: boolean;
  inputSuccess?: boolean;
  maxLength?: number;
  required?: boolean;
  showCharacterCount?: boolean;
}

/**
 * TextArea Component
 * @param {string}  id                      A unique identifier that associates the label with the TextArea.
 * @param {string}  label                   Field label text.
 * @param {string}  [name]                  Name of the textarea field. Maps to the standard HTML `name` attribute.
 * @param {string}  [characterCountMessage] Text to be displayed before the character count.
 * @param {string}  [errorMessage]          Error message text displayed when inputError === true.
 * @param {boolean} [inputError]            Triggers error message and error styling.
 * @param {boolean} [inputSuccess]          Trigger success styling.
 * @param {number}  [maxLength]             Maximum number of characters the textarea can receive.
 * @param {boolean} [required]              Adds semantic required attr and appends an * to the end of the input label.
 * @param {boolean} [showCharacterCount]    Show the character count. If maxLength is set, character count will appear as a fraction.
 */

export const TextArea = forwardRef<HTMLTextAreaElement, Props>(
  function TextArea({ onChange, ...props }, ref) {
    const {
      id,
      label,
      name,
      characterCountMessage,
      errorMessage,
      inputError = false,
      inputSuccess = false,
      maxLength,
      required = false,
      showCharacterCount = false,
      ...otherProps
    } = props;

    const [charCount, setCharCount] = useState<string>("0");

    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
      // Execute the parent form's onChange event if it exists:
      if (onChange) {
        onChange(e);
      }

      setCharCount(e.target.value.length.toString());
    };

    return (
      <div
        className={
          "usa-form-group" + (inputError ? " usa-form-group--error" : "")
        }
      >
        <label
          className={"usa-label" + (inputError ? " usa-label--error" : "")}
          htmlFor={id}
        >
          {label}
          {required && <span className="required-star">*</span>}
        </label>

        {inputError && errorMessage && (
          <span className="usa-error-message" role="alert">
            {errorMessage}
          </span>
        )}

        {showCharacterCount && (
          <span className="usa-hint">
            {characterCountMessage ? `${characterCountMessage} ` : ""}
            {charCount}
            {maxLength !== undefined && ` / ${maxLength}`}
          </span>
        )}

        <textarea
          className={
            "usa-textarea" +
            (inputError ? " usa-input-group--error" : "") +
            (inputSuccess ? " usa-input--success" : "")
          }
          id={id}
          maxLength={maxLength}
          name={name}
          onChange={handleChange}
          ref={ref}
          required={required}
          {...otherProps}
        ></textarea>
      </div>
    );
  }
);
