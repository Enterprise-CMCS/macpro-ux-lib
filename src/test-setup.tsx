import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
