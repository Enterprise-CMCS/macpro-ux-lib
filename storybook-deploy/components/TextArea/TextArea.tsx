import React, { useState } from "react";

type IntrinsicElements = JSX.IntrinsicElements["textarea"];

interface Props extends IntrinsicElements {
  characterCountMessage?: string;
  errorMessage?: string;
  fieldName: string;
  id: string;
  inputError?: boolean;
  inputFilter?: RegExp;
  inputSuccess?: boolean;
  label: string;
  maxLength?: number;
  showCharacterCount?: boolean;
  required?: boolean;
}
/**
 * TextArea Component
 * @param {string}  label                   Field label.
 * @param {string}  fieldName               Name of the input field.
 * @param {string}  id                      A unique identifier for the input.
 * @param {string}  [characterCountMessage] Sets a message preceding the character count when showCharacterCount === true.
 * @param {string}  [errorMessage]          Error message text displayed when inputError === true.
 * @param {boolean} [inputError]            Triggers error message and error styling.
 * @param {RegExp}  [inputFilter]           Used to limit input values. If a RegExp is not provided, all input types are allowed.
 * @param {boolean} [inputSuccess]          Trigger success styling.
 * @param {number}  [maxLength]             Maximum number of characters the textarea can receive.
 * @param {boolean} [showCharacterCount]    Shows the character counter. If maxlength is set, character count is shown as a fraction.
 * @param {boolean} [required]              Adds semantic required attr and appends an * to the end of the input label.
 */
export const TextArea: React.FC<Props> = ({
  label,
  fieldName,
  characterCountMessage,
  errorMessage,
  id,
  inputError = false,
  inputFilter,
  inputSuccess = false,
  maxLength,
  showCharacterCount = false,
  required = false,
  ...rest
}) => {
  const [focused, setFocused] = useState<boolean>(false);
  const [inputValue, setInputValue] = useState<string>("");

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    if (inputFilter && inputFilter.test(e.target.value)) {
      setInputValue(e.target.value);
    } else if (!inputFilter) {
      setInputValue(e.target.value);
    }
  };

  return (
    <div
      className={`usa-form-group${inputError ? " usa-form-group--error" : ""}`}
    >
      <label
        className={`usa-label${inputError ? " usa-label--error" : ""}`}
        htmlFor={`input-type-textarea-${id}`}
      >
        {label}
        {required && <span style={{ color: "#E51C3E" }}>*</span>}
      </label>
      {inputError && (
        <span
          className="usa-error-message"
          id="input-error-message"
          role="alert"
        >
          {errorMessage}
        </span>
      )}
      {showCharacterCount && (
        <span className="usa-hint">
          {characterCountMessage ? `${characterCountMessage} ` : ""}
          {inputValue.length}
          {maxLength !== undefined && ` / ${maxLength}`}
        </span>
      )}
      <textarea
        aria-describedby={`${inputError ? "input-error-message" : ""}`}
        className={`usa-textarea${inputError ? " usa-input-group--error" : ""}${
          inputSuccess ? " usa-input--success" : ""
        }${focused ? " usa-focus" : ""}`}
        id={`input-type-textarea-${id}`}
        maxLength={maxLength}
        name={fieldName}
        onBlur={() => setFocused(false)}
        onChange={handleChange}
        onFocus={() => setFocused(true)}
        required={required}
        value={inputValue}
        {...rest}
      ></textarea>
    </div>
  );
};
