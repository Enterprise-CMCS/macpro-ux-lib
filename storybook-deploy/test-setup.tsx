import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
