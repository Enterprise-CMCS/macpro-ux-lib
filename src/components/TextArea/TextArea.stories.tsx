import React from "react";
import { StoryFn, Meta } from "@storybook/react";

import { TextArea } from "./TextArea";

export default {
  title: "COMPONENTS/TextArea",
  component: TextArea,
  args: {
    label: "Input Label",
    name: "input-type-text",
    inputError: false,
    inputSuccess: false,
    required: false,
  },
  argTypes: {
    id: {
      description: "Unique ID for the input field.",
    },
    label: {
      description: "Field label.",
    },
    name: {
      description: "Name of the input field.",
    },
    characterCountMessage: {
      description: "Text to be displayed before the character count.",
    },
    errorMessage: {
      description: "Error message text displayed when inputError === true.",
    },
    inputError: {
      description: "Triggers error message and error styling.",
    },
    inputSuccess: {
      description: "Trigger success styling.",
    },
    maxLength: {
      description: "Maximum number of characters the textarea can receive.",
    },
    required: {
      description:
        "Adds semantic required attr and appends an * to the end of the input label.",
    },
    showCharacterCount: {
      description:
        "Show the character count. If maxLength is set, character count will appear as a fraction.",
    },
  },
} as Meta<typeof TextArea>;

const Template: StoryFn<typeof TextArea> = ({ ...rest }) => (
  <TextArea {...rest} />
);

export const Default = Template.bind({});
export const CharacterCount = Template.bind({});
export const Required = Template.bind({});
export const WithError = Template.bind({});
export const Success = Template.bind({});
Default.args = {};

CharacterCount.args = {
  label: "TextArea with Character Count",
  characterCountMessage: "Available remaining characters:",
  showCharacterCount: true,
  maxLength: 500,
  id: "character-count-textarea",
};

Required.args = {
  label: "Required TextArea",
  required: true,
  id: "required-textarea",
};

WithError.args = {
  errorMessage: "Helpful Error Message",
  inputError: true,
  label: "TextArea with Error Message",
  id: "required-and-error-textarea",
};

Success.args = {
  label: "Field with Success Indicator",
  initialValue: "This is a good value!",
  inputSuccess: true,
  id: "success-textarea",
};
