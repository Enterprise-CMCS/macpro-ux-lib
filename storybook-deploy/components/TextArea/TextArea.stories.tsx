import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { TextArea } from "./TextArea";

export default {
  title: "USWDS/Base/TextArea",
  component: TextArea,
  args: {
    label: "Input Label",
    fieldName: "input-type-text",
    inputError: false,
    inputSuccess: false,
    required: false,
  },
  argTypes: {
    label: {
      description: "Field label.",
    },
    fieldName: {
      description: "Name of the input field.",
    },
    characterCountMessage: {
      description:
        "Sets a message preceding the character count when showCharacterCount === true.",
    },
    errorMessage: {
      description: "Error message text displayed when inputError === true.",
    },
    inputError: {
      description: "Triggers error message and error styling.",
    },
    inputFilter: {
      description:
        "Used to limit input values. If a RegExp is not provided, all input types are allowed.",
    },
    inputSuccess: {
      description: "Trigger success styling.",
    },
    maxLength: {
      description: "Maximum number of characters the textarea can receive.",
    },
    required: {
      description:
        "Adds semantic required attr and appends an * to the end of the input label.",
    },
  },
} as ComponentMeta<typeof TextArea>;

const Template: ComponentStory<typeof TextArea> = ({ ...rest }) => (
  <TextArea {...rest} />
);

export const Default = Template.bind({});
export const CharachterCount = Template.bind({});
export const RequiredAndError = Template.bind({});
export const Success = Template.bind({});
export const InputFilter = Template.bind({});
Default.args = {};

CharachterCount.args = {
  characterCountMessage: "Available remaining characters:",
  showCharacterCount: true,
  maxLength: 500,
  id: "character-count-textarea",
};

RequiredAndError.args = {
  errorMessage: "Helpful Error Message",
  inputError: true,
  label: "Required Input Field",
  required: true,
  id: "required-and-error-textarea",
};

Success.args = {
  label: "Field with Success Indicator",
  initialValue: "This is a good value!",
  inputSuccess: true,
  id: "success-textarea",
};

InputFilter.args = {
  label: "This field only accepts a numerical input",
  inputFilter: /^-?\d*$/i,
  id: "filter-textarea",
};
