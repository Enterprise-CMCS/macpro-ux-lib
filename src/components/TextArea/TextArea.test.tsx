import React from "react";
import { screen, render, getByLabelText, getByText } from "../../test-setup";
import fireEvent from "@testing-library/user-event";
import { TextArea } from "./TextArea";

describe("TextArea component", () => {
  describe("default", () => {
    const errorMessage = "My name is Tom Riddle";
    const characterCountMessage = "This has a counter";

    beforeEach(() => {
      render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
          errorMessage={errorMessage}
        />
      );
    });

    it("should render", () => {
      const comp = screen.getByLabelText("Testing TextArea");
      expect(comp).toBeInTheDocument();
    });

    it("should not have an initial value", () => {
      const comp = screen.getByLabelText("Testing TextArea");
      expect(comp).toHaveDisplayValue("");
    });

    it("should accept any value", () => {
      const comp = screen.getByLabelText("Testing TextArea");
      fireEvent.type(comp, "abc123!@#");
      expect(comp).toHaveDisplayValue("abc123!@#");
    });

    it("should not show the errorMessage", () => {
      const comp = screen.getByLabelText("Testing TextArea");
      expect(comp).toBeInTheDocument();
      expect(screen.queryAllByText(errorMessage).length).toBe(0);
    });

    it("should not show the characterCountMessage", () => {
      const comp = screen.getByLabelText("Testing TextArea");
      expect(comp).toBeInTheDocument();
      expect(screen.queryAllByText(characterCountMessage).length).toBe(0);
    });
  });

  describe("charachter counter", () => {
    it("should render with the character counter", () => {
      const { container } = render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
          showCharacterCount
        />
      );
      const comp = screen.getByLabelText("Testing TextArea");
      fireEvent.type(comp, "four");
      expect(comp).toHaveDisplayValue("four");
      const counter = container.getElementsByClassName("usa-hint")[0];
      expect(counter.innerHTML).toBe("4");
    });

    it("should render character counter as a fraction", () => {
      const { container } = render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
          showCharacterCount
          maxLength={50}
        />
      );
      const comp = screen.getByLabelText("Testing TextArea");
      fireEvent.type(comp, "four");
      expect(comp).toHaveDisplayValue("four");
      const counter = container.getElementsByClassName("usa-hint")[0];
      expect(counter.innerHTML).toBe("4 / 50");
    });

    it("should render character counter with message", () => {
      const { container } = render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
          characterCountMessage="You have used:"
          showCharacterCount
          maxLength={50}
        />
      );
      const comp = screen.getByLabelText("Testing TextArea");
      fireEvent.type(comp, "four");
      expect(comp).toHaveDisplayValue("four");
      const counter = container.getElementsByClassName("usa-hint")[0];
      expect(counter.innerHTML).toBe("You have used: 4 / 50");
    });
  });

  it("should add 'usa-focus' on focus'", () => {
    const { container } = render(
      <TextArea
        id="test-input-area"
        label="Testing Input"
        fieldName="testing-input"
        errorMessage="Click me to unfocus input"
        inputError
        required
      />
    );
    // Should not exist on render
    const textArea = getByLabelText(container, "Testing Input*");
    expect(textArea.className.includes("usa-focus")).toBe(false);

    // Should exist onFocus
    fireEvent.click(textArea);
    expect(textArea.className.includes("usa-focus")).toBe(true);

    // Should not onBlur
    const escape = getByText(container, "Click me to unfocus input");
    fireEvent.click(escape);
    expect(textArea.className.includes("usa-focus")).toBe(false);
  });

  it("should show an error message", () => {
    const errorMessage = "My name is Tom Riddle";
    render(
      <TextArea
        id="test-input-area"
        label="Testing TextArea"
        fieldName="testing-textarea"
        errorMessage={errorMessage}
        inputError
      />
    );
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it("should be required", () => {
    render(
      <TextArea
        id="test-input-area"
        label="Testing TextArea"
        fieldName="testing-textarea"
        required
      />
    );
    const comp = screen.getByLabelText("Testing TextArea*");
    expect(comp).toBeInTheDocument();
    comp.hasAttribute("required");
  });

  it("should accept a numeric value only", () => {
    const filter = new RegExp(/^-?\d*$/i);
    render(
      <TextArea
        id="test-input-area"
        label="Testing TextArea"
        fieldName="testing-textarea"
        inputFilter={filter}
      />
    );
    const comp = screen.getByLabelText("Testing TextArea");
    fireEvent.type(comp, "abc123");
    expect(comp).toHaveDisplayValue("123");
  });

  describe("compontent snapshots", () => {
    it("default", () => {
      const { container } = render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
        />
      );
      expect(container).toMatchSnapshot();
    });
    it("required and error", () => {
      const errorMessage = "My name is Tom Riddle";
      const { container } = render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
          errorMessage={errorMessage}
          required
          placeholder="Can you place this?"
        />
      );
      expect(container).toMatchSnapshot();
    });
    it("success", () => {
      const { container } = render(
        <TextArea
          id="test-input-area"
          label="Testing TextArea"
          fieldName="testing-textarea"
          inputSuccess
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
