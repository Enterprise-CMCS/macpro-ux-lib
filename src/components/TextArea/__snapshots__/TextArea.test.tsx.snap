// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea component compontent snapshots renders a component with an error message 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      name="testing-textarea"
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders a required component 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
      <span
        class="required-star"
      >
        *
      </span>
    </label>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      name="testing-textarea"
      required=""
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders the default component 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      name="testing-textarea"
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders with a character counter 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <span
      class="usa-hint"
    >
      0
    </span>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      name="testing-textarea"
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders with a character counter and maxLength 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <span
      class="usa-hint"
    >
      0
       / 50
    </span>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      maxlength="50"
      name="testing-textarea"
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders with a character counter and message 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <span
      class="usa-hint"
    >
      This is a character count message 
      0
    </span>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      name="testing-textarea"
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders with a character counter, maxLength, and message 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <span
      class="usa-hint"
    >
      This is a character count message 
      0
       / 50
    </span>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      maxlength="50"
      name="testing-textarea"
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots renders with a placeholder 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <textarea
      class="usa-textarea"
      id="test-input-area"
      name="testing-textarea"
      placeholder="This is example placeholder text."
    />
  </div>
</div>
`;

exports[`TextArea component compontent snapshots success 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-input-area"
    >
      Testing TextArea
    </label>
    <textarea
      class="usa-textarea usa-input--success"
      id="test-input-area"
      name="testing-textarea"
    />
  </div>
</div>
`;
